README.md
# wlroots

Pluggable, composable, unopinionated modules for building a [Wayland]
compositor; or about 60,000 lines of code you were going to write anyway.

- wlroots provides backends that abstract the underlying display and input
  hardware, including KMS/DRM, libinput, Wayland, X11, and headless backends,
  plus any custom backends you choose to write, which can all be created or
  destroyed at runtime and used in concert with each other.
- wlroots provides unopinionated, mostly standalone implementations of many
  Wayland interfaces, both from wayland.xml and various protocol extensions.
  We also promote the standardization of portable extensions across
  many compositors.
- wlroots provides several powerful, standalone, and optional tools that
  implement components common to many compositors, such as the arrangement of
  outputs in physical space.
- wlroots provides an Xwayland abstraction that allows you to have excellent
  Xwayland support without worrying about writing your own X11 window manager
  on top of writing your compositor.
- wlroots provides a renderer abstraction that simple compositors can use to
  avoid writing GL code directly, but which steps out of the way when your
  needs demand custom rendering code.

wlroots implements a huge variety of Wayland compositor features and implements
them *right*, so you can focus on the features that make your compositor
unique. By using wlroots, you get high performance, excellent hardware
compatibility, broad support for many wayland interfaces, and comfortable
development tools - or any subset of these features you like, because all of
them work independently of one another and freely compose with anything you want
to implement yourself.

Check out our [wiki] to get started with wlroots. Join our IRC channel:
[#sway-devel on Libera Chat].

A variety of [wrapper libraries] are available for using it with your favorite
programming language.

## Building

Install dependencies:

* meson
* wayland
* wayland-protocols
* EGL and GLESv2 (optional, for the GLES2 renderer)
* Vulkan loader, headers and glslang (optional, for the Vulkan renderer)
* libdrm
* GBM (optional, for the GBM allocator)
* libinput (optional, for the libinput backend)
* xkbcommon
* udev
* pixman
* [libseat]
* [hwdata] (optional, for the DRM backend)
* [libliftoff-rpi] (optional, for the DRM backend)

If you choose to enable X11 support:

* xwayland (build-time only, optional at runtime)
* libxcb
* libxcb-render-util
* libxcb-wm
* libxcb-errors (optional, for improved error reporting)

Run these commands:

    meson build/
    ninja -C build/

Install like so:

    sudo ninja -C build/ install

## Contributing

See [CONTRIBUTING.md].

[Wayland]: https://wayland.freedesktop.org/
[wiki]: https://gitlab.freedesktop.org/wlroots/wlroots/-/wikis/Getting-started
[#sway-devel on Libera Chat]: https://web.libera.chat/gamja/?channels=#sway-devel
[wrapper libraries]: https://gitlab.freedesktop.org/wlroots/wlroots/-/wikis/Projects-which-use-wlroots#wrapper-libraries
[libseat]: https://git.sr.ht/~kennylevinsen/seatd
[hwdata]: https://github.com/vcrhonek/hwdata
[CONTRIBUTING.md]: https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/master/CONTRIBUTING.md


debian/changelog
wlroots (0.16.0-2~bpo11+1~rpt1) bullseye; urgency=medium

  * Backport from experimental

 -- Simon Long <simon@raspberrypi.com>  Thu, 01 Dec 2022 16:32:04 +0000

wlroots (0.16.0-2) experimental; urgency=medium

  * [17dcb12] Fix compilation due to uninitialized values on some
    architectures.

 -- Guido Günther <agx@sigxcpu.org>  Thu, 24 Nov 2022 13:50:49 +0100

wlroots (0.16.0-1) experimental; urgency=medium

  * Upload to experimental
  * [aac440c] New upstream release
  * [e8cf5bd] d/gbp.conf: Use upstream-vcs
  * [bea6239] d/watch: Project moved to fdo gitlab
  * [e7033ad] Adjust to new soname
  * [b9ef4e6] d/control: Update build-dependencies
  * [5054262] Refresh patches. Drop those applied upstream.

 -- Guido Günther <agx@sigxcpu.org>  Wed, 23 Nov 2022 16:20:46 +0100

wlroots (0.15.1-3) unstable; urgency=medium

  * [01320b4] Backport upstream patch to fix build with ffmpeg 5.0
    (Closes: #1004792)

 -- Guido Günther <agx@sigxcpu.org>  Fri, 24 Jun 2022 11:54:33 +0200

wlroots (0.15.1-2) unstable; urgency=medium

  * [d2f0630] xdg-activation: Deduplicate token creation code
    Backport a (not yet merged) MR to fix a crash in the
    xdg-activation code.

 -- Guido Günther <agx@sigxcpu.org>  Sat, 26 Mar 2022 19:34:50 +0100

wlroots (0.15.1-1) unstable; urgency=medium

  * Upload to unstable
  * d/control: Fix typo (Closes: #984583)
  * d/copyright: Fix source URL
  * d/control: Fix homepage URL.
    Project moved to gitlab (Closes: #1001154)
  * New upstream version (Closes: #1005892)

 -- Guido Günther <agx@sigxcpu.org>  Fri, 25 Mar 2022 11:57:45 +0100

wlroots (0.15.0-3) experimental; urgency=medium

  * [f909925] d/control: Bump meson and wayland-protocol dependencies.
    Adjust to what the project requires (Closes: #966176)

 -- Guido Günther <agx@sigxcpu.org>  Fri, 18 Mar 2022 15:11:32 +0100

wlroots (0.15.0-2) experimental; urgency=medium

  * [7c01dd7] Use libdrm 2.4.109.
    This allows us to drop the reverted upstream change.
  * [1784c6e] Use wayland 1.20.
    This allows us to revert the reverted upstream change
  * [f60b70a] Bump standards version
  * [838be17] d/copyright: Drop unused pattern

 -- Guido Günther <agx@sigxcpu.org>  Mon, 14 Feb 2022 15:25:46 +0100

wlroots (0.15.0-1) experimental; urgency=medium

  * New upstream release
  * [bb15cc9] d/control: Bump build-deps according to meson dependencies
  * [bf804d6] Rediff patches
  * [c551f31] d/rules: Enable all optional backends
  * [1d55195] Adjust to soname bump
  * Use headless backend in tests
  * Revert several patches so we can build against libdrm and libwayland
    currently in Debian.

 -- Guido Günther <agx@sigxcpu.org>  Fri, 17 Dec 2021 15:15:52 +0100

wlroots (0.14.1-5) unstable; urgency=medium

  * Upload to unstable
  * [865f2f1] Drop patches backported form upstream

 -- Guido Günther <agx@sigxcpu.org>  Thu, 16 Dec 2021 09:39:07 +0100

wlroots (0.14.1-4) experimental; urgency=medium

  [ Birger Schacht ]
  * [3fafe04] Add x11 backend build dependencies also to the -dev package

 -- Guido Günther <agx@sigxcpu.org>  Mon, 08 Nov 2021 18:53:18 +0100

wlroots (0.14.1-3) experimental; urgency=medium

  * [59c2af5] d/control: Add more dependencies for x11 backend.
    This reenables the x11 backend
  * [d1c7991] d/rules: Force x11 backend.
    This makes sure it doesn't get disabled again due to lack of
    dependencies

 -- Guido Günther <agx@sigxcpu.org>  Sat, 06 Nov 2021 11:42:15 +0100

wlroots (0.14.1-2) experimental; urgency=medium

  [ Birger Schacht ]
  * [d512fe8] Enable XWayland again

  [ Guido Günther ]
  * [9f1d30d] d/control: Build-depend on libseat-dev

 -- Guido Günther <agx@sigxcpu.org>  Fri, 05 Nov 2021 21:37:46 +0100

wlroots (0.14.1-1) experimental; urgency=medium

  * New upstream version 0.14.1
  * Bump soname to 9
  * d/control: Build-depend on libseat-dev

 -- Guido Günther <agx@sigxcpu.org>  Mon, 27 Sep 2021 22:36:50 +0200

wlroots (0.13.0-1) experimental; urgency=medium

  * New upstream version 0.13.0
  * Adust to soname change
  * d/control: Bump libwayland-dev to 0.19
  * autopkgtest: Adjust to API change

 -- Guido Günther <agx@sigxcpu.org>  Tue, 13 Apr 2021 11:02:16 +0200

wlroots (0.12.0-3) unstable; urgency=medium

  * [02be191] examples/dmabuf-capture: stop using av_init_packet
    (Closes: #996500)

 -- Guido Günther <agx@sigxcpu.org>  Tue, 26 Oct 2021 10:13:01 +0200

wlroots (0.12.0-2) unstable; urgency=medium

  [ nicoo ]
  * Team upload
  * Kick off to unstable

  [ Léo Andrès ]
  * [2524851] libwlroots0: fix typo in description.

 -- nicoo <nicoo@debian.org>  Sun, 14 Feb 2021 12:31:16 +0100

wlroots (0.12.0-1) experimental; urgency=medium

  * Upload to experimental
  * [45e00a5] Drop patches. All applied upstream
  * [ee0f2eb] Revert "layer-shell: error on 0 dimension without anchors"
    Undo this until phosh has a fixed release.
  * [6866955] d/gbp.conf: Use xz when generating tarballs
  * [788b999] debian: Adjust to soname change
  * [32602cb] d/control: Bump standards version
  * [efa76a2] d/upstream/metadata: Add upstream bug tracker URLs
  * [2ac8de7] d/rules: Enable all hardening opts

 -- Guido Günther <agx@sigxcpu.org>  Fri, 01 Jan 2021 15:29:35 +0100

wlroots (0.11.0-3) unstable; urgency=medium

  * [c95d2e9] output-management-v1: send complete head state on enable change.
    Fixes a phosh crash when the compositor fails to send updated
    state.

 -- Guido Günther <agx@sigxcpu.org>  Mon, 12 Oct 2020 12:40:11 +0200

wlroots (0.11.0-2) unstable; urgency=medium

  * Upload to unstable
  * [0b2ae69] libwlroots-examples: Install more examples
  * [da6f836] Build-depend on libav*-dev.
    This allows to build the screencopy-dmabuf example

 -- Guido Günther <agx@sigxcpu.org>  Fri, 24 Jul 2020 10:22:21 +0200

wlroots (0.11.0-1) experimental; urgency=medium

  * Upload to experimental
  * New upstream version 0.11.0
    (Closes: #957935)
  * Drop all patches - all are part of 0.11.0
  * Adjust to soname change
  * Update d/copyright
  * Switch to dh 13


backend/drm/util.c
#define _POSIX_C_SOURCE 200809L
#include <assert.h>
#include <drm_fourcc.h>
#include <drm_mode.h>
#include <drm.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <wlr/util/log.h>
#include "backend/drm/drm.h"
#include "backend/drm/util.h"

int32_t calculate_refresh_rate(const drmModeModeInfo *mode) {
	int32_t refresh = (mode->clock * 1000000LL / mode->htotal +
		mode->vtotal / 2) / mode->vtotal;

	if (mode->flags & DRM_MODE_FLAG_INTERLACE) {
		refresh *= 2;
	}

	if (mode->flags & DRM_MODE_FLAG_DBLSCAN) {
		refresh /= 2;
	}

	if (mode->vscan > 1) {
		refresh /= mode->vscan;
	}

	return refresh;
}

enum wlr_output_mode_aspect_ratio get_picture_aspect_ratio(const drmModeModeInfo *mode) {
	switch (mode->flags & DRM_MODE_FLAG_PIC_AR_MASK) {
	case DRM_MODE_FLAG_PIC_AR_NONE:
		return WLR_OUTPUT_MODE_ASPECT_RATIO_NONE;
	case DRM_MODE_FLAG_PIC_AR_4_3:
		return WLR_OUTPUT_MODE_ASPECT_RATIO_4_3;
	case DRM_MODE_FLAG_PIC_AR_16_9:
		return WLR_OUTPUT_MODE_ASPECT_RATIO_16_9;
	case DRM_MODE_FLAG_PIC_AR_64_27:
		return WLR_OUTPUT_MODE_ASPECT_RATIO_64_27;
	case DRM_MODE_FLAG_PIC_AR_256_135:
		return WLR_OUTPUT_MODE_ASPECT_RATIO_256_135;
	default:
		wlr_log(WLR_ERROR, "Unknown mode picture aspect ratio: %u",
			mode->flags & DRM_MODE_FLAG_PIC_AR_MASK);
		return WLR_OUTPUT_MODE_ASPECT_RATIO_NONE;
	}
}

/* See https://en.wikipedia.org/wiki/Extended_Display_Identification_Data for layout of EDID data.
 * We don't parse the EDID properly. We just expect to receive valid data.
 */
void parse_edid(struct wlr_drm_connector *conn, size_t len, const uint8_t *data) {
	struct wlr_output *output = &conn->output;

	free(output->make);
	free(output->model);
	free(output->serial);
	output->make = NULL;
	output->model = NULL;
	output->serial = NULL;

	if (!data || len < 128) {
		return;
	}

	uint16_t id = (data[8] << 8) | data[9];
	const char *manu = get_pnp_manufacturer(id);
	char pnp_id[4];
	if (!manu) {
		// The ASCII 3-letter manufacturer PnP ID is encoded in 5-bit codes
		pnp_id[0] = ((id >> 10) & 0x1F) + '@';
		pnp_id[1] = ((id >> 5) & 0x1F) + '@';
		pnp_id[2] = ((id >> 0) & 0x1F) + '@';
		pnp_id[3] = '\0';
		manu = pnp_id;
	}
	output->make = strdup(manu);

	uint16_t model = data[10] | (data[11] << 8);
	char model_str[32];
	snprintf(model_str, sizeof(model_str), "0x%04" PRIX16, model);

	uint32_t serial = data[12] | (data[13] << 8) | (data[14] << 8) | (data[15] << 8);
	char serial_str[32];
	if (serial != 0) {
		snprintf(serial_str, sizeof(serial_str), "0x%08" PRIX32, serial);
	} else {
		serial_str[0] = '\0';
	}

	for (size_t i = 72; i <= 108; i += 18) {
		uint16_t flag = (data[i] << 8) | data[i + 1];
		if (flag == 0 && data[i + 3] == 0xFC) {
			snprintf(model_str, sizeof(model_str), "%.13s", &data[i + 5]);

			// Monitor names are terminated by newline if they're too short
			char *nl = strchr(model_str, '\n');
			if (nl) {
				*nl = '\0';
			}
		} else if (flag == 0 && data[i + 3] == 0xFF) {
			snprintf(serial_str, sizeof(serial_str), "%.13s", &data[i + 5]);

			// Monitor serial numbers are terminated by newline if they're too
			// short
			char* nl = strchr(serial_str, '\n');

			if (nl) {
				*nl = '\0';
			}
		}
	}

	output->model = strdup(model_str);
	if (serial_str[0] != '\0') {
		output->serial = strdup(serial_str);
	}
}

const char *drm_connector_status_str(drmModeConnection status)
{
   switch (status)
     {
      case DRM_MODE_CONNECTED:
        return "connected";
      case DRM_MODE_DISCONNECTED:
        return "disconnected";
      case DRM_MODE_UNKNOWNCONNECTION:
        return "unknown";
     }

   return "<unsupported>";
}

static bool is_taken(size_t n, const uint32_t arr[static n], uint32_t key) {
	for (size_t i = 0; i < n; ++i) {
		if (arr[i] == key) {
			return true;
		}
	}
	return false;
}

/*
 * Store all of the non-recursive state in a struct, so we aren't literally
 * passing 12 arguments to a function.
 */
struct match_state {
	const size_t num_objs;
	const uint32_t *restrict objs;
	const size_t num_res;
	size_t score;
	size_t replaced;
	uint32_t *restrict res;
	uint32_t *restrict best;
	const uint32_t *restrict orig;
	bool exit_early;
};

/*
 * skips: The number of SKIP elements encountered so far.
 * score: The number of resources we've matched so far.
 * replaced: The number of changes from the original solution.
 * i: The index of the current element.
 *
 * This tries to match a solution as close to st->orig as it can.
 *
 * Returns whether we've set a new best element with this solution.
 */
static bool match_obj_(struct match_state *st, size_t skips, size_t score, size_t replaced, size_t i) {
	// Finished
	if (i >= st->num_res) {
		if (score > st->score ||
				(score == st->score && replaced < st->replaced)) {
			st->score = score;
			st->replaced = replaced;
			memcpy(st->best, st->res, sizeof(st->best[0]) * st->num_res);

			st->exit_early = (st->score == st->num_res - skips
					|| st->score == st->num_objs)
					&& st->replaced == 0;

			return true;
		} else {
			return false;
		}
	}

	if (st->orig[i] == SKIP) {
		st->res[i] = SKIP;
		return match_obj_(st, skips + 1, score, replaced, i + 1);
	}

	bool has_best = false;

	/*
	 * Attempt to use the current solution first, to try and avoid
	 * recalculating everything
	 */
	if (st->orig[i] != UNMATCHED && !is_taken(i, st->res, st->orig[i])) {
		st->res[i] = st->orig[i];
		size_t obj_score = st->objs[st->res[i]] != 0 ? 1 : 0;
		if (match_obj_(st, skips, score + obj_score, replaced, i + 1)) {
			has_best = true;
		}
	}
	if (st->orig[i] == UNMATCHED) {
		st->res[i] = UNMATCHED;
		if (match_obj_(st, skips, score, replaced, i + 1)) {
			has_best = true;
		}
	}
	if (st->exit_early) {
		return true;
	}

	if (st->orig[i] != UNMATCHED) {
		++replaced;
	}

	for (size_t candidate = 0; candidate < st->num_objs; ++candidate) {
		// We tried this earlier
		if (candidate == st->orig[i]) {
			continue;
		}

		// Not compatible
		if (!(st->objs[candidate] & (1 << i))) {
			continue;
		}

		// Already taken
		if (is_taken(i, st->res, candidate)) {
			continue;
		}

		st->res[i] = candidate;
		size_t obj_score = st->objs[candidate] != 0 ? 1 : 0;
		if (match_obj_(st, skips, score + obj_score, replaced, i + 1)) {
			has_best = true;
		}

		if (st->exit_early) {
			return true;
		}
	}

	if (has_best) {
		return true;
	}

	// Maybe this resource can't be matched
	st->res[i] = UNMATCHED;
	return match_obj_(st, skips, score, replaced, i + 1);
}

size_t match_obj(size_t num_objs, const uint32_t objs[static restrict num_objs],
		size_t num_res, const uint32_t res[static restrict num_res],
		uint32_t out[static restrict num_res]) {
	uint32_t solution[num_res];
	for (size_t i = 0; i < num_res; ++i) {
		solution[i] = UNMATCHED;
	}

	struct match_state st = {
		.num_objs = num_objs,
		.num_res = num_res,
		.score = 0,
		.replaced = SIZE_MAX,
		.objs = objs,
		.res = solution,
		.best = out,
		.orig = res,
		.exit_early = false,
	};

	match_obj_(&st, 0, 0, 0, 0);
	return st.score;
}
